lalegpl/lib/lale/nsga2.py

```diff
@@ -0,0 +1,567 @@
+        if not all_type_same:
+            varg["variator"] = CompoundOperator(SBX(), HUX(), PM(), BitFlip())
+
+        algorithm = NSGAII(
+            problem,
+            population_size=self.population_size,
+            **varg,
+        )
+
+        try:
+            algorithm.run(adjusted_max_evals, callback=time_check_callback)
+        except MaxBudgetExceededException as e:
+            logger.warning(
+                f"Max optimization time budget exceeded. Optimization exited prematurely.\n{e}"
+            )
+
+        solutions = nondominated(algorithm.result)
+        # solutions = [s for s in algorithm.result if s.feasible]`
+        # solutions = algorithm.result
+
+        moo_solutions = []
+        for solution in solutions:
+            vars = []
+            for pnum in range(parameter_num):
+                vars.append(problem.types[pnum].decode(solution.variables[pnum]))
+
+            vars_dict = self.param_to_dict(
+                vars,
+                self.model_helper.param_choices,
+                self.model_helper.param_categories,
+                self.model_helper.param_type,
+            )
+            moo_solutions.append(self.Soln(vars_dict, solution.objectives))
+            logger.info(f"{vars}, {solution.objectives}")
+
+        self.moo_solutions = moo_solutions
+
+        pareto_models = []
+        for solution in self.moo_solutions:
+            est = self.model_helper.create_instance(solution.variables)
+            est_trained = est.fit(X, y)
+            pareto_models.append((solution.variables, est_trained))
+
+        self.pareto_models = pareto_models
+        return self
+
+    def get_pareto_solutions(self):
+        return self.moo_solutions
+
+    def get_pareto_models(self):
+        return self.pareto_models
+
+    # Predict using first pareto-optimal estimator
+    def predict(self, X, **kwargs):
+        if "pipeline_name" in kwargs:
+            pname = kwargs["pipeline_name"]
+            pipeline = self.get_pipeline(pipeline_name=pname)
+            del kwargs["pipeline_name"]
+        else:
+            pipeline = self.get_pipeline()
+
+        return pipeline.predict(X, **kwargs)
+
+    # Return pareto-optimal estimator
+    def get_pipeline(self, pipeline_name=None, astype="lale"):
+        """Retrieve one of the pareto-optimal pipelines.
+
+        Parameters
+        ----------
+        pipeline_name : union type, default None
+
+            - string
+                Key (name) from the table returned by summary(), return a trained pipeline.
+
+            - None
+                When not specified, return the first (trained) pipeline in the table
+                returned by summary()
+
+        astype : 'lale' or 'sklearn', default 'lale'
+            Type of resulting pipeline.
+
+        Returns
+        -------
+        result : Trained operator."""
+
+        id = 0
+        if pipeline_name is not None:
+            id = int(pipeline_name[1:])
```

Comment:
Minor:
IMO, this appears to be a somewhat non-robust way of getting `id` from `pipeline_name`. Might be worthwhile to have a `(name, pipeline)` dictionary or some guidance to the user as to how the pipeline names should be set.